File: ezr/0-atom-c/scope_test.c
/*
 * scope_test.c — Ever / Tapestry, universal map verification
 *
 * Tests the hash function, insert/lookup/delete, resize, scope chain,
 * all five entry kinds, and the wire round-trip through EValue(RECORD).
 */

#include <stdio.h>
#include <string.h>
#include <math.h>
#include "scope.h"

static int ok_n = 0, fail_n = 0;
static void _ok(const char *name, int cond) {
    if (cond) { ok_n++;   printf("  \u2713 %s\n", name); }
    else       { fail_n++; printf("  \u2717 %s\n", name); }
}
#define ok(n, c) _ok(n, c)

int main(void) {
    ev_pool  pool;    ev_pool_init(&pool);
    EvArena *arena = ev_arena_new("scope-test");

    printf("\n=== Ever \u2014 universal map (EvScope) ===\n\n");

    /* ── FNV-1a hash ── */
    printf("FNV-1a hash\n");
    ok("hash non-zero",         ev_fnv1a("x") != 0);
    ok("hash deterministic",    ev_fnv1a("hello") == ev_fnv1a("hello"));
    ok("hash distinct a/b",     ev_fnv1a("abc") != ev_fnv1a("def"));
    ok("hash empty string",     ev_fnv1a("") != 0);   /* FNV offset basis */
    ok("hash long name",        ev_fnv1a("my_very_long_module_name") != 0);

    /* ── Lifecycle ── */
    printf("\nLifecycle\n");
    EvScope *global = ev_scope_new(arena, NULL, "global");
    ok("global created",        global != NULL);
    ok("global depth 0",        global->depth == 0);
    ok("global cap 64",         global->cap == EV_SCOPE_INIT_CAP);
    ok("global used 0",         global->used == 0);
    ok("global parent null",    global->parent == NULL);
    ok("global name set",       strcmp(global->name, "global") == 0);

    /* ── Insert and lookup — VAR ── */
    printf("\nInsert and lookup — VAR\n");
    ok("set x=42",       ev_scope_set_var(global, "x", ev_int(42),
                                           E_CERTAIN, E_LANG_EVER));
    ok("set y=3.14",     ev_scope_set_var(global, "y", ev_real(3.14),
                                           E_CERTAIN, E_LANG_PYTHON));
    ok("set flag=true",  ev_scope_set_var(global, "flag", ev_bool(1),
                                           E_CERTAIN, E_LANG_EVER));
    ok("set name=Codric",ev_scope_set_var(global, "name", ev_text("Codric"),
                                           E_CERTAIN, E_LANG_EVER));

    EvScopeEntry *ex = ev_scope_get(global, "x");
    ok("get x != NULL",        ex != NULL);
    ok("x value = 42",         ex && ex->value.body.as_int == 42);
    ok("x kind = VAR",         ex && ex->kind == EV_SK_VAR);
    ok("x confidence = 256",   ex && ex->confidence == E_CERTAIN);
    ok("x lang = EVER",        ex && ex->lang == E_LANG_EVER);

    EvScopeEntry *ey = ev_scope_get(global, "y");
    ok("get y != NULL",        ey != NULL);
    ok("y value ≈ 3.14",       ey && fabs(ey->value.body.as_real - 3.14) < 1e-9);
    ok("y lang = PYTHON",      ey && ey->lang == E_LANG_PYTHON);

    ok("get missing = NULL",   ev_scope_get(global, "zzz") == NULL);
    ok("lookup missing = NULL", ev_scope_lookup(global, "zzz") == NULL);
    ok("used = 4",             global->used == 4);

    /* ── Update existing ── */
    printf("\nUpdate existing binding\n");
    ok("update x=99",    ev_scope_set_var(global, "x", ev_int(99),
                                           200, E_LANG_EVER));
    EvScopeEntry *ex2 = ev_scope_get(global, "x");
    ok("x updated to 99",      ex2 && ex2->value.body.as_int == 99);
    ok("used still 4",         global->used == 4);  /* update, not insert */

    /* ── Function entries ── */
    printf("\nFunction entries — EV_SK_FN\n");
    EvArena *fa = ev_arena_new("fn-arena");
    EvNode *body = ev_node_binop(fa, EV_NODE_MUL,
                                 ev_node_var(fa,"n",1),
                                 ev_node_var(fa,"n",1), 1);
    const char *prms[] = {"n"};
    ok("set_fn square",  ev_scope_set_fn(global, "square", body,
                                          prms, 1, E_CERTAIN, E_LANG_EVER));

    EvScopeEntry *efn = ev_scope_get(global, "square");
    ok("get square",           efn != NULL);
    ok("square kind = FN",     efn && efn->kind == EV_SK_FN);
    ok("square node set",      efn && efn->node == body);
    ok("square param_count=1", efn && efn->param_count == 1);
    ok("square param name",    efn && strcmp(efn->params[0], "n") == 0);

    EvNode *got_fn = ev_scope_get_fn(global, "square");
    ok("get_fn returns body",  got_fn == body);
    ok("get_fn missing=NULL",  ev_scope_get_fn(global, "nope") == NULL);
    ev_arena_free(fa);

    /* ── Module entries ── */
    printf("\nModule entries — EV_SK_MODULE\n");
    EvArena *ma = ev_arena_new("math-arena");
    EvScope *math_mod = ev_scope_new(ma, global, "Math");
    ev_scope_set_var(math_mod, "PI", ev_real(3.14159265358979),
                     E_CERTAIN, E_LANG_EVER);
    ev_scope_set_module(global, "Math", math_mod, E_LANG_EVER);

    EvScopeEntry *em = ev_scope_get(global, "Math");
    ok("get Math module",      em != NULL);
    ok("Math kind = MODULE",   em && em->kind == EV_SK_MODULE);

    EvScope *got_mod = ev_scope_get_module(global, "Math");
    ok("get_module returns scope", got_mod == math_mod);
    ok("module PI = π",
       fabs(ev_scope_get_value(got_mod,"PI").body.as_real - 3.14159265) < 1e-7);
    ev_arena_free(ma);

    /* ── Scope chain — local inherits from global ── */
    printf("\nScope chain — local inherits global\n");
    EvArena *la = ev_arena_new("local-arena");
    EvScope *local = ev_scope_new(la, global, "fn:add");
    ok("local depth = 1",      local->depth == 1);
    ok("local parent = global",local->parent == global);

    ev_scope_set_var(local, "a", ev_int(10), E_CERTAIN, E_LANG_EVER);
    ev_scope_set_var(local, "b", ev_int(32), E_CERTAIN, E_LANG_EVER);

    /* local binding found in local */
    EvScopeEntry *ea = ev_scope_lookup(local, "a");
    ok("lookup 'a' in local",  ea && ea->value.body.as_int == 10);

    /* global binding found by walking chain */
    EvScopeEntry *ex3 = ev_scope_lookup(local, "x");
    ok("lookup 'x' via chain", ex3 && ex3->value.body.as_int == 99);

    /* shadow: local 'x' shadows global 'x' */
    ev_scope_set_var(local, "x", ev_int(0), E_CERTAIN, E_LANG_EVER);
    EvScopeEntry *shadow = ev_scope_lookup(local, "x");
    ok("shadow x = 0 (local)",  shadow && shadow->value.body.as_int == 0);

    /* global x unchanged */
    EvScopeEntry *global_x = ev_scope_get(global, "x");
    ok("global x still = 99",  global_x && global_x->value.body.as_int == 99);

    /* get_value walks chain */
    EValue bval = ev_scope_get_value(local, "b");
    ok("get_value 'b'",         bval.body.as_int == 32);
    EValue missing = ev_scope_get_value(local, "zzz");
    ok("get_value missing=void", missing.tag == EV_VOID);
    ev_arena_free(la);

    /* ── Deletion / tombstones ── */
    printf("\nDeletion (tombstones)\n");
    EvArena *da = ev_arena_new("del-arena");
    EvScope *ds = ev_scope_new(da, NULL, "del-test");
    ev_scope_set_var(ds, "a", ev_int(1), E_CERTAIN, E_LANG_EVER);
    ev_scope_set_var(ds, "b", ev_int(2), E_CERTAIN, E_LANG_EVER);
    ev_scope_set_var(ds, "c", ev_int(3), E_CERTAIN, E_LANG_EVER);

    ok("delete 'b'",            ev_scope_delete(ds, "b"));
    ok("get 'b' = NULL",        ev_scope_get(ds, "b") == NULL);
    ok("get 'a' still live",    ev_scope_get(ds, "a") != NULL);
    ok("get 'c' still live",    ev_scope_get(ds, "c") != NULL);
    ok("delete missing = 0",    !ev_scope_delete(ds, "zzz"));
    ok("used decremented",      ds->used == 2);

    /* re-insert after deletion */
    ev_scope_set_var(ds, "b", ev_int(99), E_CERTAIN, E_LANG_EVER);
    ok("re-insert 'b'=99",      ev_scope_get(ds,"b") != NULL);
    ok("re-inserted value",
       ev_scope_get(ds,"b")->value.body.as_int == 99);
    ev_scope_free(ds); ev_arena_free(da);

    /* ── Resize — force growth past 75 % ── */
    printf("\nResize — force growth past 75%%\n");
    EvArena *ra = ev_arena_new("resize-arena");
    EvScope *rs = ev_scope_new(ra, NULL, "resize");
    char key[16];
    /* insert 50 items — exceeds 64 * 0.75 = 48 threshold */
    for (int i = 0; i < 50; i++) {
        snprintf(key, sizeof key, "var_%d", i);
        ev_scope_set_var(rs, key, ev_int(i), E_CERTAIN, E_LANG_EVER);
    }
    ok("50 items inserted",     rs->used == 50);
    ok("cap grew past 64",      rs->cap > EV_SCOPE_INIT_CAP);

    /* all 50 items still retrievable after resize */
    int all_ok = 1;
    for (int i = 0; i < 50; i++) {
        snprintf(key, sizeof key, "var_%d", i);
        EvScopeEntry *e = ev_scope_get(rs, key);
        if (!e || e->value.body.as_int != i) { all_ok = 0; break; }
    }
    ok("all 50 items survive resize", all_ok);
    ev_scope_free(rs); ev_arena_free(ra);

    /* ── Wire format — scope ↔ EValue(RECORD) ── */
    printf("\nWire format — scope \u2194 EValue(RECORD)\n");
    EvArena *wa = ev_arena_new("wire-arena");
    EvScope *ws = ev_scope_new(wa, NULL, "wire");
    ev_scope_set_var(ws, "id",    ev_int(7),         E_CERTAIN, E_LANG_SQL);
    ev_scope_set_var(ws, "name",  ev_text("Alice"),   E_CERTAIN, E_LANG_SQL);
    ev_scope_set_var(ws, "score", ev_real(98.6),      E_CERTAIN, E_LANG_SQL);

    EValue rec = ev_scope_to_record(ws, &pool);
    ok("to_record tag = RECORD",  rec.tag == EV_RECORD);

    EvScope *ws2 = ev_scope_new(wa, NULL, "wire2");
    ok("from_record",             ev_scope_from_record(ws2, &rec, &pool,
                                                        E_LANG_SQL));
    ok("id round-trips",
       ev_scope_get_value(ws2,"id").body.as_int == 7);
    ok("name round-trips",
       strcmp(ev_scope_get_value(ws2,"name").text, "Alice") == 0);
    ok("score round-trips",
       fabs(ev_scope_get_value(ws2,"score").body.as_real - 98.6) < 1e-9);

    /* serialise the RECORD further through EValue wire format */
    uint8_t buf[2048];
    int32_t n = ev_serialise(&rec, buf, sizeof buf, &pool);
    ok("EValue serialise succeeds", n > 0);
    int32_t consumed = 0;
    EValue rec2 = ev_deserialise(buf, n, &consumed, &pool);
    ok("EValue round-trip",       rec2.tag == EV_RECORD && consumed == n);

    EvScope *ws3 = ev_scope_new(wa, NULL, "wire3");
    ev_scope_from_record(ws3, &rec2, &pool, E_LANG_PYTHON);
    ok("C→wire→Python→C id",
       ev_scope_get_value(ws3,"id").body.as_int == 7);
    ev_scope_free(ws);  ev_scope_free(ws2);  ev_scope_free(ws3);
    ev_arena_free(wa);

    /* ── Stats and dump ── */
    printf("\nDiagnostics\n");
    char stats[128];
    ev_scope_stats(global, stats, sizeof stats);
    ok("stats non-empty",       strlen(stats) > 10);
    ok("stats has scope name",  strstr(stats, "global") != NULL);
    printf("  %s\n", stats);

    /* ── Cleanup ── */
    ev_scope_free(global);
    ev_arena_free(arena);

    printf("\n=== EvScope: %d passed, %d failed ===\n\n", ok_n, fail_n);
    return fail_n == 0 ? 0 : 1;
}

File: ezr/0-atom-c/2-interpreter-python/scope_test.py
#!/usr/bin/env python3
"""scope_test.py — Ever / Tapestry, Python scope verification."""

import math
from ir import parse_to_ir, EvNode, NK
from scope import (EvScope, ScopeEntry, SK, fnv1a,
                    make_global_scope, eval_node)
from evalue import EValue, EvType, EV_CERTAIN, serialise, deserialise

p = f = 0
def ok(n, c):
    global p, f
    if c: p += 1; print(f"  \u2713 {n}")
    else: f += 1; print(f"  \u2717 {n}")

print("\n=== EvScope — Python universal map ===\n")

print("FNV-1a hash (matches scope.c)")
ok("fnv1a non-zero",        fnv1a("x") != 0)
ok("fnv1a deterministic",   fnv1a("hello") == fnv1a("hello"))
ok("fnv1a distinct",        fnv1a("abc") != fnv1a("def"))
ok("fnv1a empty",           fnv1a("") != 0)
# Exact value match with C implementation (FNV-1a 32-bit)
ok("fnv1a('x') == C value", fnv1a("x") == 0xfd0c5087)

print("\nLifecycle")
g = EvScope(name="global")
ok("created",          g is not None)
ok("depth 0",          g.depth == 0)
ok("len 0",            len(g) == 0)
ok("parent None",      g.parent is None)
ok("name set",         g.name == "global")

print("\nInsert and lookup — VAR")
g.set_var("x",     EValue.from_int(42))
g.set_var("y",     EValue.from_real(3.14))
g.set_var("flag",  EValue.from_bool(True))
g.set_var("name",  EValue.from_text("Codric"))

e = g.get("x")
ok("get x",           e is not None)
ok("x value = 42",    e and e.value.ev_int == 42)
ok("x kind = VAR",    e and e.kind == SK.VAR)
ok("x conf = 256",    e and e.confidence == EV_CERTAIN)

ok("get y real",      g.get("y") and abs(g.get("y").value.ev_real - 3.14) < 1e-9)
ok("get flag bool",   g.get("flag") and g.get("flag").value.ev_bool is True)
ok("get name text",   g.get("name") and g.get("name").value.ev_text == "Codric")
ok("get missing",     g.get("zzz") is None)
ok("lookup missing",  g.lookup("zzz") is None)
ok("len = 4",         len(g) == 4)

print("\nUpdate")
g.set_var("x", EValue.from_int(99))
ok("x updated to 99", g.get("x").value.ev_int == 99)
ok("len still 4",     len(g) == 4)

print("\nFunction entries — SK.FN")
body = EvNode.binop("*", EvNode.var("n"), EvNode.var("n"))
g.set_fn("square", body, ["n"])
e_fn = g.get("square")
ok("get square",      e_fn is not None)
ok("kind = FN",       e_fn and e_fn.kind == SK.FN)
ok("node = body",     e_fn and e_fn.node is body)
ok("params = ['n']",  e_fn and e_fn.params == ["n"])
ok("get_fn returns",  g.get_fn("square") is body)
ok("get_fn missing",  g.get_fn("nope") is None)

print("\nModule entries — SK.MODULE")
math_mod = EvScope(parent=g, name="Math")
math_mod.set_var("PI", EValue.from_real(math.pi))
g.set_module("Math", math_mod)
e_m = g.get("Math")
ok("get Math",        e_m is not None)
ok("kind = MODULE",   e_m and e_m.kind == SK.MODULE)
ok("get_module",      g.get_module("Math") is math_mod)
ok("module PI",       abs(g.get_module("Math").get_value("PI").ev_real - math.pi) < 1e-9)

print("\nScope chain — local inherits global")
local = EvScope(parent=g, name="fn:add")
ok("depth = 1",       local.depth == 1)
ok("parent = global", local.parent is g)
local.set_var("a", EValue.from_int(10))
local.set_var("b", EValue.from_int(32))
ok("lookup a local",  local.lookup("a").value.ev_int == 10)
ok("lookup x chain",  local.lookup("x").value.ev_int == 99)
local.set_var("x", EValue.from_int(0))
ok("shadow x = 0",    local.lookup("x").value.ev_int == 0)
ok("global x = 99",   g.get("x").value.ev_int == 99)
ok("get_value b",     local.get_value("b").ev_int == 32)
ok("get_value void",  local.get_value("zzz").ev_tag == EvType.VOID)

print("\nDelete")
ds = EvScope(name="del")
ds.set_var("a", EValue.from_int(1))
ds.set_var("b", EValue.from_int(2))
ds.set_var("c", EValue.from_int(3))
ok("delete b",        ds.delete("b"))
ok("b gone",          ds.get("b") is None)
ok("a still live",    ds.get("a") is not None)
ok("delete missing",  not ds.delete("zzz"))
ok("len = 2",         len(ds) == 2)
ds.set_var("b", EValue.from_int(99))
ok("re-insert b=99",  ds.get("b").value.ev_int == 99)

print("\nBulk insert (100 entries)")
bulk = EvScope(name="bulk")
for i in range(100):
    bulk.set_var(f"var_{i}", EValue.from_int(i))
ok("100 inserted",    len(bulk) == 100)
all_ok = all(bulk.get(f"var_{i}") and
              bulk.get(f"var_{i}").value.ev_int == i
              for i in range(100))
ok("all 100 correct", all_ok)

print("\nWire format — scope ↔ EValue(RECORD)")
ws = EvScope(name="wire")
ws.set_var("id",    EValue.from_int(7))
ws.set_var("name",  EValue.from_text("Alice"))
ws.set_var("score", EValue.from_real(98.6))
rec = ws.to_record()
ok("to_record RECORD",   rec.ev_tag == EvType.RECORD)
ok("to_record has id",   rec.get("id") is not None and rec.get("id").ev_int == 7)
ok("to_record has name", rec.get("name") and rec.get("name").ev_text == "Alice")
ws2 = EvScope.from_record(rec, name="wire2")
ok("from_record id",     ws2.get_value("id").ev_int == 7)
ok("from_record name",   ws2.get_value("name").ev_text == "Alice")
ok("from_record score",  abs(ws2.get_value("score").ev_real - 98.6) < 1e-9)
# serialise through evalue wire format
buf = serialise(rec)
back_rec, pos = deserialise(buf)
ok("EValue wire serialise",  pos == len(buf))
ws3 = EvScope.from_record(back_rec, name="wire3")
ok("C→wire→Python id",   ws3.get_value("id").ev_int == 7)

print("\neval_node through scope")
ev_g = make_global_scope()
ev_g.set_var("x", EValue.from_int(6))
ev_g.set_var("y", EValue.from_int(36))

for src, expected, label in [
    ("6 + 36",                 42,  "literal add"),
    ("if true then 42 else 0", 42,  "if true"),
    ("if false then 0 else 42",42,  "if false"),
]:
    r = parse_to_ir(src)
    ok(f"parse {label}", r.ok)
    if r.ok:
        result = eval_node(r.root, EvScope(parent=ev_g))
        ok(f"eval {label} = {expected}",
           result.ev_tag == EvType.INT and result.ev_int == expected)

# define and call a function through the scope
fact_scope = EvScope(parent=ev_g, name="fact-test")
r2 = parse_to_ir("def fact(n) = if n <= 1 then 1 else n * fact(n - 1)")
ok("parse fact", r2.ok)
eval_node(r2.root, fact_scope)  # registers fact
ok("fact registered",  "fact" in fact_scope)

for n_val, expected in [(1,1),(2,2),(3,6)]:
    call = EvNode.fn_call("fact", [EvNode.int_lit(n_val)])
    result = eval_node(call, fact_scope, depth=0)
    ok(f"fact({n_val}) = {expected}",
       result.ev_tag == EvType.INT and result.ev_int == expected)

# depth ceiling enforced
call4 = EvNode.fn_call("fact", [EvNode.int_lit(4)])
r4 = eval_node(call4, fact_scope, depth=0)
ok("fact(4) void (depth ceiling)", r4.ev_tag == EvType.VOID)

print("\nBuiltin scope")
bg = make_global_scope()
ok("true pre-loaded",  bg.get("true") and bg.get("true").value.ev_bool is True)
ok("false pre-loaded", bg.get("false") and bg.get("false").value.ev_bool is False)
ok("z pre-loaded",     bg.get("z") and bg.get("z").value.ev_tag == EvType.VOID)
ok("print builtin",    bg.get("print") and bg.get("print").kind == SK.BUILTIN)

print(f"\n=== EvScope Python: {p} passed, {f} failed ===\n")
raise SystemExit(0 if f == 0 else 1)
